sdk/schemaregistry/schema-registry-avro/test/schemaRegistryAvroSerializer.spec.ts

```diff
@@ -0,0 +1,211 @@
+    const payload = testAvroType.toBuffer(testValue);
+    const buffer = Buffer.alloc(36 + payload.length);
+
+    buffer.write(testSchemaIds[0], 4, 32, "utf-8");
+    payload.copy(buffer, 36);
+    assertSameJsonRepresentation(await serializer.deserialize(buffer), testValue);
+  });
+
+  it("serializes and deserializes in round trip", async () => {
+    let serializer = createTestSerializer();
+    let buffer = await serializer.serialize(testValue, testSchema);
+    assertSameJsonRepresentation(await serializer.deserialize(buffer), testValue);
+
+    // again for cache hit coverage on serialize
+    buffer = await serializer.serialize(testValue, testSchema);
+    assertSameJsonRepresentation(await serializer.deserialize(buffer), testValue);
+
+    // throw away serializer for cache miss coverage on deserialize
+    serializer = createTestSerializer(false);
+    assertSameJsonRepresentation(await serializer.deserialize(buffer), testValue);
+
+    // thow away serializer again and cover getSchemaId instead of registerSchema
+    serializer = createTestSerializer(false);
+    assert.deepStrictEqual(await serializer.serialize(testValue, testSchema), buffer);
+  });
+});
+
+function assertSameJsonRepresentation(actual: any, expected: any) {
+  // avsc returns objects with different prototype than our plain object literal
+  // so assert.deepEqual fails.
+  //
+  // REVIEW: Is it a problem that this avsc detail leaks?
+  assert.strictEqual(JSON.stringify(actual), JSON.stringify(expected));
+}
+
+function createTestSerializer(autoRegisterSchemas = true): SchemaRegistryAvroSerializer {
+  const client = createTestRegistry();
+
+  if (!autoRegisterSchemas) {
+    client.registerSchema({
+      name: `${testSchemaObject.namespace}.${testSchemaObject.name}`,
+      group: "TestGroup",
+      content: testSchema,
+      serializationType: "avro"
+    });
+  }
+
+  return new SchemaRegistryAvroSerializer(client, "TestGroup", { autoRegisterSchemas });
+}
+
+function createTestRegistry(): SchemaRegistry {
+  const mapById = new Map<string, Schema>();
+  const mapByContent = new Map<string, Schema>();
+  let idCounter = 0;
+
+  return { registerSchema, getSchemaId, getSchemaById };
+
+  async function registerSchema(
+    schema: SchemaDescription,
+    _options?: RegisterSchemaOptions
+  ): Promise<SchemaId> {
+    let result = mapByContent.get(schema.content);
+    if (!result) {
+      result = {
+        id: newId(),
+        content: schema.content,
+        version: 1,
+        location: "",
+        locationById: "",
+        serializationType: schema.serializationType
+      };
+      mapByContent.set(result.content, result);
+      mapById.set(result.id, result);
+    }
+    return result;
+
+    function newId(): string {
+      if (idCounter == testSchemaIds.length) {
+        throw new Error("Out of IDs. Generate more GUIDs and paste them above.");
+      }
+      const id = testSchemaIds[idCounter];
+      idCounter++;
```

Comment:
I applaud you for resisting the temptation to write `return testSchemaIds[idCounter++];` -- I think breaking it out like this is much more understandable to the casual reader.

Reply:
Thanks. The temptation was strong. :)